src/attachment.rs

```diff
@@ -0,0 +1,344 @@
+#[no_mangle]
+pub extern "C" fn z_bytes_map_drop(this: &mut z_owned_bytes_map_t) {
+    let this = unsafe { &mut *this.get() };
+    this.take();
+}
+
+/// Returns the value associated with `key`, returning a gravestone value if:
+/// - `this` or `key` is in gravestone state.
+/// - `this` has no value associated to `key`
+#[no_mangle]
+pub extern "C" fn z_bytes_map_get(this: &z_owned_bytes_map_t, key: z_bytes_t) -> z_bytes_t {
+    let this = unsafe { &*this.get() };
+    let (Some(this), Some(key)) = (this.as_ref(), key.as_slice()) else {
+        return z_bytes_null();
+    };
+    if let Some(value) = this.get(key) {
+        value.as_ref().into()
+    } else {
+        z_bytes_null()
+    }
+}
+
+/// Associates `value` to `key` in the map, copying them to obtain ownership: `key` and `value` are not aliased past the function's return.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_insert_by_copy(
+    this: &z_owned_bytes_map_t,
+    key: z_bytes_t,
+    value: z_bytes_t,
+) {
+    let this = unsafe { &mut *this.get() };
+    if let (Some(this), Some(key), Some(value)) = (this.as_mut(), key.as_slice(), value.as_slice())
+    {
+        this.insert(Cow::Owned(key.to_owned()), Cow::Owned(value.to_owned()));
+    }
+}
+
+/// Associates `value` to `key` in the map, aliasing them.
+///
+/// Note that once `key` is aliased, reinserting at the same key may alias the previous instance, or the new instance of `key`.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_insert_by_alias(
+    this: &z_owned_bytes_map_t,
+    key: z_bytes_t,
+    value: z_bytes_t,
+) {
+    let this = unsafe { &mut *this.get() };
+    if let (Some(this), Some(key), Some(value)) = (this.as_mut(), key.as_slice(), value.as_slice())
+    {
+        unsafe {
+            this.insert(
+                Cow::Borrowed(core::mem::transmute(key)),
+                Cow::Borrowed(core::mem::transmute(value)),
+            )
+        };
+    }
+}
+
+/// Iterates over the key-value pairs in the map.
+///
+/// `body` will be called once per pair, with `ctx` as its last argument.
+/// If `body` returns a non-zero value, the iteration will stop immediately and the value will be returned.
+/// Otherwise, this will return 0 once all pairs have been visited.
+/// `body` is not given ownership of the key nor value, which alias the pairs in the map.
+/// It is safe to keep these aliases until existing keys are modified/removed, or the map is destroyed.
+/// Note that this map is unordered.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_iter(
+    this: &z_owned_bytes_map_t,
+    body: z_attachment_iter_body_t,
+    ctx: *mut c_void,
+) -> i8 {
+    let this = unsafe { &*this.get() };
```

Comment:
The `Deref` trait should be defined for `z_owned_bytes_map_t`, like here https://github.com/eclipse-zenoh/zenoh-c/blob/master/src/keyexpr.rs#L75

Then it will be possible to write `let this = this.as_ref()` without repearting `unsafe`:
https://github.com/eclipse-zenoh/zenoh-c/blob/master/src/keyexpr.rs#L139

Reply:
As discussed, will leave it as is for now.